### PR TITLE
Initial integration tests with BCFIPS distribution

### LINKED ISSUE #14886: Initial integration tests with BCFIPS distribution

### PR DESCRIPTION
**Pull request by mposolda**
Closes #14886

PR introduces support for testing with BCFIPS based Keycloak Quarkus distribution.

Summary of changes:
- File `ci.yml` contains one new group for testing with BCFIPS based Keycloak distribution. So far, it tests only package `org.keycloak.testsuite.forms`. We can consider more packages in the future (possibly not whole testsuite, but just stuff interesting from "crypto" perspective). I've tried to run whole testsuite on the RHEL FIPS host - there are still some failing tests (related to X509 auth, Kerberos and SAML). I am planning to address them in a follow-up issue

- Unit tests as well as integration tests are using custom `java.security` file with only BC FIPS security providers included. So it doesn't test with the security providers from the default `java.security` file. This allows to test with same security providers on the non-FIPS JDK on GH actions as well as on FIPS JDK on RHEL FIPS. Another reason is, that default security providers from RHEL FIPS OpenJDK still contain security providers, which are not 100% FIPS compliant and hence with including them, it is not 100% confident that Keycloak uses only FIPS certified stuff - details in https://access.redhat.com/documentation/en-us/openjdk/11/html-single/configuring_openjdk_11_on_rhel_with_fips/index#ref_openjdk-default-fips-configuration_openjdk

- The above point requires that BC FIPS Jars need to be available on bootstrap classpath of Keycloak quarkus Java process. As the security providers are bootstrapped as very 1st step of JVM run (before quarkus bootstrap). So added small tweak to `kc.sh`/`kc.bat` to include JARs from `lib/bootstrap` directory to bootstrap classpath (if this directory exists)

- Small tweaks in the quarkus distribution. Especially make sure that `CryptoIntegration` is initialized before `KeycloakSessionFactory` initialization, which is needed to make sure that Keycloak SPIs are initialized with correct security providers

- Small tweaks in the crypto integration. Especially make sure that security providers are added during initialization of `CryptoProvider` itself rather then with `BouncyIntegration` . This is due the fact that `FIPS1402Provider` needs to add more security providers (there is also BCJSSE provider in addition to BCFIPS provider for have TLS working)

- Updated `docs/fips.md` with instructions on how to run the distribution. Contains options for:
-- Including BCFIPS jars in `providers` directory. In this case they are not required on bootstrap classpath as system is still using also default security providers from `java.security` file.
-- Including BCFIPS jars in the `lib/bootstrap` directory and use custom `java.security` file to include only BCFIPS security providers. This is the option used also for the automated tests

### CODE REVIEW

**Comment by mposolda** [comment]
@stianst Do you please have opportunity to review this?

**Comment by stianst** [comment]
@pedroigor is finally back, which is good as he should be able to give a better review than me

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -274,6 +274,86 @@ jobs:
...
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 2
+
+      - name: Check whether HEAD^ contains crypto related changes
+        run: echo "GIT_CRYPTO_RELEVANT_DIFF=$( git diff --name-only HEAD^ | egrep -ic -e 'crypto' )" >> $GITHUB_ENV
+
+      - name: Cache Maven packages
+        if: ${{ github.event_name != 'pull_request' || env.GIT_CRYPTO_RELEVANT_DIFF != 0 }}
```
Comment (pedroigor) [inline .github/workflows/ci.yml:295]: I think I know the reasoning behind the condition. However, given how FIPS changes the runtime and how much the server rely on crypto operations, wouldn't be safer to always run the tests?

File: quarkus/deployment/src/main/java/org/keycloak/quarkus/deployment/KeycloakProcessor.java [hunk 2]
```diff
@@ -305,7 +306,7 @@ private void configureDefaultPersistenceUnitEntities(ParsedPersistenceXmlDescrip
      *
      * @param recorder
      */
-    @Consume(BootstrapConfigSetupCompleteBuildItem.class)
+    @Consume(CryptoProviderInitBuildItem.class)
     @Record(ExecutionTime.STATIC_INIT)
     @BuildStep
     KeycloakSessionFactoryPreInitBuildItem configureProviders(KeycloakRecorder recorder, List<PersistenceXmlDescriptorBuildItem> descriptors) {
```
Comment (pedroigor) [inline quarkus/deployment/src/main/java/org/keycloak/quarkus/deployment/KeycloakProcessor.java:312]: I would avoid creating a built item here and change this method name to `configureKeycloakSessionFactory` to make sure the provider is set before bootstrapping the session factory.

I don't see a reason for creating another built item if its purpose is only for ordering.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (mposolda) [reply .github/workflows/ci.yml]: Yes, I was a bit undecided whether to add condition or not and I've added it first. But I've removed it now based on your feedback. We can see if GH action builds are OK and eventually add the condition later if it turns that PR builds are too long...

File: quarkus/deployment/src/main/java/org/keycloak/quarkus/deployment/KeycloakProcessor.java (hunk 2 above)
Comment (mposolda) [reply quarkus/deployment/src/main/java/org/keycloak/quarkus/deployment/KeycloakProcessor.java]: Yes, I confirm the only point for introducing build item was ordering. I've changed as you suggested. It works as expected and cryptoProvider is initialized before keycloak session factory. Thanks for this tip!

**Review by pedroigor — APPROVED** [review]
@mposolda Thanks.

**Comment by mposolda** [comment]
@pedroigor Thanks to you as well

**Pull request merged**